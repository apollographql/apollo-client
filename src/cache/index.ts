export { Transaction, ApolloCache } from './core/cache';
export { Cache } from './core/types/Cache';
export { DataProxy } from './core/types/DataProxy';
export { MissingFieldError } from './core/types/common';

export {
  Reference,
  isReference,
  makeReference,
} from '../utilities/graphql/storeUtils';

export {
  InMemoryCache,
  InMemoryCacheConfig,
  ReactiveVar,
} from './inmemory/inMemoryCache';

export {
  defaultDataIdFromObject,
  TypePolicies,
  TypePolicy,
  FieldPolicy,
  FieldReadFunction,
  FieldMergeFunction,
  FieldFunctionOptions,
  PossibleTypesMap,
} from './inmemory/policies';

export * from './inmemory/types';
