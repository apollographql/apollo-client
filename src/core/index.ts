/* Core */

export {
  ApolloClient,
  ApolloClientOptions,
  DefaultOptions,
  mergeOptions,
} from './ApolloClient';
export {
  ObservableQuery,
  FetchMoreOptions,
  UpdateQueryOptions,
} from './ObservableQuery';
export {
  QueryOptions,
  WatchQueryOptions,
  MutationOptions,
  SubscriptionOptions,
  FetchPolicy,
  WatchQueryFetchPolicy,
  ErrorPolicy,
  FetchMoreQueryOptions,
  SubscribeToMoreOptions,
  MutationUpdaterFn,
} from './watchQueryOptions';
export { NetworkStatus } from './networkStatus';
export * from './types';
export {
  Resolver,
  FragmentMatcher,
} from './LocalState';
export { isApolloError, ApolloError } from '../errors';

/* Cache */

export {
  // All the exports (types and values) from ../cache, minus cacheSlot,
  // which we want to keep semi-private.
  Cache,
  ApolloCache,
  Transaction,
  DataProxy,
  InMemoryCache,
  InMemoryCacheConfig,
  MissingFieldError,
  defaultDataIdFromObject,
  ReactiveVar,
  makeVar,
  TypePolicies,
  TypePolicy,
  FieldPolicy,
  FieldReadFunction,
  FieldMergeFunction,
  FieldFunctionOptions,
  PossibleTypesMap,
} from '../cache';

export * from '../cache/inmemory/types';

/* Link */

export * from '../link/core';
export * from '../link/http';
export {
  fromError,
  toPromise,
  fromPromise,
  ServerError,
  throwServerError,
} from '../link/utils';

/* Utilities */

export {
  Observable,
  Observer,
  ObservableSubscription,
  Reference,
  isReference,
  makeReference,
  StoreObject,
} from '../utilities';

/* Supporting */

// The verbosity of invariant.{log,warn,error} can be controlled globally
// (for anyone using the same ts-invariant package) by passing "log",
// "warn", "error", or "silent" to setVerbosity. By default, Apollo Client
// displays warnings and errors, but hides invariant.log statements. Note
// that all invariant.* logging is hidden in production.
import { setVerbosity } from "ts-invariant";
export { setVerbosity as setLogVerbosity }
setVerbosity("warn");

// Note that importing `gql` by itself, then destructuring
// additional properties separately before exporting, is intentional.
// Due to the way the `graphql-tag` library is setup, certain bundlers
// can't find the properties added to the exported `gql` function without
// additional guidance (e.g. Rollup - see
// https://rollupjs.org/guide/en/#error-name-is-not-exported-by-module).
// Instead of having people that are using bundlers with `@apollo/client` add
// extra bundler config to help `graphql-tag` exports be found (which would be
// awkward since they aren't importing `graphql-tag` themselves), this
// workaround of pulling the extra properties off the `gql` function,
// then re-exporting them separately, helps keeps bundlers happy without any
// additional config changes.
import gql from 'graphql-tag';
export const {
  resetCaches,
  disableFragmentWarnings,
  enableExperimentalFragmentVariables,
  disableExperimentalFragmentVariables
} = gql;
export { gql };
