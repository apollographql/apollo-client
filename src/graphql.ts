/**
 * This type represents the dynamic form of any data that may be returned from
 * a GraphQL API. Basically this is just a narrower `mixed` type, or another
 * type for JSON.
 */
export type GraphQLData = null | boolean | number | string | GraphQLArrayData | GraphQLObjectData;

/**
 * An that may be returned from a GraphQL API.
 */
export interface GraphQLObjectData { [fieldName: string]: GraphQLData; }

/**
 * A list of values that may be returned from a GraphQL API.
 */
export interface GraphQLArrayData extends Array<GraphQLData> {}

/**
 * A guard function which checks to see if the provided value is GraphQL object
 * data. True if it is.
 */
export function isObjectData (value: any): value is GraphQLObjectData {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}
