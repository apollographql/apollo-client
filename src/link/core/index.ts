export { empty } from './empty';
export { from } from './from';
export { split } from './split';
export { concat } from './concat';
export { execute } from './execute';
export { ApolloLink } from './ApolloLink';

export * from './types';
