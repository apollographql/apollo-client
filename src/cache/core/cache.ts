import { DocumentNode } from 'graphql';
import { wrap } from 'optimism';

import { getFragmentQueryDocument } from '../../utilities/graphql/fragments';
import { StoreObject } from '../../utilities/graphql/storeUtils';
import { DataProxy } from './types/DataProxy';
import { Cache } from './types/Cache';
import { Modifier, Modifiers } from './types/common';

export type Transaction<T> = (c: ApolloCache<T>) => void;

export abstract class ApolloCache<TSerialized> implements DataProxy {
  // required to implement
  // core API
  public abstract read<T, TVariables = any>(
    query: Cache.ReadOptions<TVariables>,
  ): T | null;
  public abstract write<TResult = any, TVariables = any>(
    write: Cache.WriteOptions<TResult, TVariables>,
  ): void;
  public abstract diff<T>(query: Cache.DiffOptions): Cache.DiffResult<T>;
  public abstract watch(watch: Cache.WatchOptions): () => void;
  public abstract reset(): Promise<void>;

  // If called with only one argument, removes the entire entity
  // identified by dataId. If called with a fieldName as well, removes all
  // fields of the identified entity whose store names match fieldName.
  public abstract evict(dataId: string, fieldName?: string): boolean;

  // intializer / offline / ssr API
  /**
   * Replaces existing state in the cache (if any) with the values expressed by
   * `serializedState`.
   *
   * Called when hydrating a cache (server side rendering, or offline storage),
   * and also (potentially) during hot reloads.
   */
  public abstract restore(
    serializedState: TSerialized,
  ): ApolloCache<TSerialized>;

  /**
   * Exposes the cache's complete state, in a serializable format for later restoration.
   */
  public abstract extract(optimistic?: boolean): TSerialized;

  // Optimistic API

  public abstract removeOptimistic(id: string): void;

  // Transactional API

  public abstract performTransaction(
    transaction: Transaction<TSerialized>,
  ): void;

  public abstract recordOptimisticTransaction(
    transaction: Transaction<TSerialized>,
    id: string,
  ): void;

  // Optional API

  public transformDocument(document: DocumentNode): DocumentNode {
    return document;
  }

  public identify(object: StoreObject): string | undefined {
    return;
  }

  public modify(
    modifiers: Modifier<any> | Modifiers,
    dataId?: string,
    optimistic?: boolean,
  ): boolean {
    return false;
  }
  
  public gc(): string[] {
    return [];
  }

  // Experimental API

  public transformForLink(document: DocumentNode): DocumentNode {
    return document;
  }

  // DataProxy API
  /**
   *
   * @param options
   * @param optimistic
   */
  public readQuery<QueryType, TVariables = any>(
    options: DataProxy.Query<TVariables>,
    optimistic: boolean = false,
  ): QueryType | null {
    return this.read({
      rootId: options.id || 'ROOT_QUERY',
      query: options.query,
      variables: options.variables,
      optimistic,
    });
  }

  // Make sure we compute the same (===) fragment query document every
  // time we receive the same fragment in readFragment.
  private getFragmentDoc = wrap(getFragmentQueryDocument);

  public readFragment<FragmentType, TVariables = any>(
    options: DataProxy.Fragment<TVariables>,
    optimistic: boolean = false,
  ): FragmentType | null {
    return this.read({
      query: this.getFragmentDoc(options.fragment, options.fragmentName),
      variables: options.variables,
      rootId: options.id,
      optimistic,
    });
  }

  public writeQuery<TData = any, TVariables = any>(
    options: Cache.WriteQueryOptions<TData, TVariables>,
  ): void {
    this.write({
      dataId: options.id || 'ROOT_QUERY',
      result: options.data,
      query: options.query,
      variables: options.variables,
    });
  }

  public writeFragment<TData = any, TVariables = any>(
    options: Cache.WriteFragmentOptions<TData, TVariables>,
  ): void {
    this.write({
      dataId: options.id,
      result: options.data,
      variables: options.variables,
      query: this.getFragmentDoc(options.fragment, options.fragmentName),
    });
  }
}
