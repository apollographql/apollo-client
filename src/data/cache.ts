import { DocumentNode } from 'graphql';
import {
  DataProxy,
  DataProxyReadQueryOptions,
  DataProxyReadFragmentOptions,
  DataProxyWriteQueryOptions,
  DataProxyWriteFragmentOptions,
} from './proxy';
import { addTypenameToDocument } from '../queries/queryTransform';
import {
  getFragmentQueryDocument,
  getOperationName,
} from '../queries/getFromAST';

export type CacheWrite = {
  dataId: string;
  result: any;
  document: DocumentNode;
  variables?: Object;
};

export abstract class Cache implements DataProxy {
  private addTypename: boolean;
  constructor(addTypename: boolean) {
    this.addTypename = addTypename;
  }

  public abstract reset(): Promise<void>;

  public abstract diffQuery(query: {
    query: DocumentNode;
    variables: any;
    returnPartialData?: boolean;
    previousResult?: any;
    optimistic: boolean;
  }): any;

  public abstract read(query: {
    query: DocumentNode;
    variables: any;
    rootId?: string;
    previousResult?: any;
    optimistic: boolean;
  }): any;

  public readQuery<QueryType>(
    options: DataProxyReadQueryOptions,
    optimistic: boolean = false,
  ): QueryType {
    let query = options.query;
    if (this.addTypename) {
      query = addTypenameToDocument(query);
    }

    return this.read({
      query,
      variables: options.variables,
      optimistic,
    });
  }

  public readFragment<FragmentType>(
    options: DataProxyReadFragmentOptions,
    optimistic: boolean = false,
  ): FragmentType | null {
    let document = getFragmentQueryDocument(
      options.fragment,
      options.fragmentName,
    );
    if (this.addTypename) {
      document = addTypenameToDocument(document);
    }

    return this.read({
      query: document,
      variables: options.variables,
      rootId: options.id,
      optimistic,
    });
  }

  public abstract writeResult(write: CacheWrite): void;

  public writeQuery(options: DataProxyWriteQueryOptions): void {
    let query = options.query;
    if (this.addTypename) {
      query = addTypenameToDocument(query);
    }

    this.writeResult({
      dataId: 'ROOT_QUERY',
      result: options.data,
      document: query,
      variables: options.variables,
    });
  }

  public writeFragment(options: DataProxyWriteFragmentOptions): void {
    let document = getFragmentQueryDocument(
      options.fragment,
      options.fragmentName,
    );
    if (this.addTypename) {
      document = addTypenameToDocument(document);
    }

    this.writeResult({
      dataId: options.id,
      result: options.data,
      document,
      variables: options.variables,
    });
  }

  public abstract removeOptimistic(id: string): void;

  public abstract performTransaction(transaction: (c: Cache) => void): void;
  public abstract recordOptimisticTransaction(
    transaction: (c: Cache) => void,
    id: string,
  ): void;
}
