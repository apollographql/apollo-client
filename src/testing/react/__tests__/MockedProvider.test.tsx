import React from 'react';
import { DocumentNode } from 'graphql';
import { render, waitFor } from '@testing-library/react';
import gql from 'graphql-tag';

import { itAsync, MockedResponse, MockLink } from '../../core';
import { MockedProvider } from '../MockedProvider';
import { useQuery } from '../../../react/hooks';
import { InMemoryCache } from '../../../cache';
import { ApolloLink } from '../../../link/core';

const variables = {
  username: 'mock_username'
};

const userWithoutTypeName = {
  id: 'user_id'
};

const user = {
  __typename: 'User',
  ...userWithoutTypeName
};

const query: DocumentNode = gql`
  query GetUser($username: String!) {
    user(username: $username) {
      id
    }
  }
`;

const queryWithTypename: DocumentNode = gql`
  query GetUser($username: String!) {
    user(username: $username) {
      id
      __typename
    }
  }
`;

const mocks: ReadonlyArray<MockedResponse> = [
  {
    request: {
      query,
      variables
    },
    result: { data: { user } }
  }
];

interface Data {
  user: {
    id: string;
  };
}

interface Variables {
  username: string;
}

let errorThrown = false;
const errorLink = new ApolloLink((operation, forward) => {
  let observer = null;
  try {
    observer = forward(operation);
  } catch (error) {
    errorThrown = true;
  }
  return observer;
});

describe('General use', () => {
  beforeEach(() => {
    errorThrown = false;
  });

  itAsync('should mock the data', (resolve, reject) => {
    let finished = false;
    function Component({ username }: Variables) {
      const { loading, data } = useQuery<Data, Variables>(query, { variables });
      if (!loading) {
        expect(data!.user).toMatchSnapshot();
        finished = true;
      }
      return null;
    }

    render(
      <MockedProvider mocks={mocks}>
        <Component {...variables} />
      </MockedProvider>
    );

    waitFor(() => {
      expect(finished).toBe(true);
    }).then(resolve, reject);
  });

  itAsync('should allow querying with the typename', (resolve, reject) => {
    let finished = false;
    function Component({ username }: Variables) {
      const { loading, data } = useQuery<Data, Variables>(query, { variables });
      if (!loading) {
        expect(data!.user).toMatchSnapshot();
        finished = true;
      }
      return null;
    }

    const mocksWithTypename = [
      {
        request: {
          query: queryWithTypename,
          variables
        },
        result: { data: { user } }
      }
    ];

    render(
      <MockedProvider mocks={mocksWithTypename}>
        <Component {...variables} />
      </MockedProvider>
    );

    waitFor(() => {
      expect(finished).toBe(true);
    }).then(resolve, reject);
  });

  itAsync('should allow using a custom cache', (resolve, reject) => {
    let finished = false;
    const cache = new InMemoryCache();
    cache.writeQuery({
      query,
      variables,
      data: { user }
    });

    function Component({ username }: Variables) {
      const { loading, data } = useQuery<Data, Variables>(query, { variables });
      if (!loading) {
        expect(data).toMatchObject({ user });
        finished = true;
      }
      return null;
    }

    render(
      <MockedProvider mocks={[]} cache={cache}>
        <Component {...variables} />
      </MockedProvider>
    );

    waitFor(() => {
      expect(finished).toBe(true);
    }).then(resolve, reject);
  });

  itAsync('should error if the variables in the mock and component do not match', (resolve, reject) => {
    let finished = false;
    function Component({ ...variables }: Variables) {
      const { loading, error } = useQuery<Data, Variables>(query, { variables });
      if (!loading) {
        expect(error).toMatchSnapshot();
        finished = true;
      }
      return null;
    }

    const variables2 = {
      username: 'other_user',
      age: undefined
    };

    render(
      <MockedProvider showWarnings={false} mocks={mocks}>
        <Component {...variables2} />
      </MockedProvider>
    );

    waitFor(() => {
      expect(finished).toBe(true);
    }).then(resolve, reject);
  });

  itAsync('should error if the variables do not deep equal', (resolve, reject) => {
    let finished = false
    function Component({ ...variables }: Variables) {
      const { loading, error } = useQuery<Data, Variables>(query, { variables });
      if (!loading) {
        expect(error).toMatchSnapshot();
        finished = true;
      }
      return null;
    }

    const mocks2 = [
      {
        request: {
          query,
          variables: {
            age: 13,
            username: 'some_user'
          }
        },
        result: { data: { user } }
      }
    ];

    const variables2 = {
      username: 'some_user',
      age: 42
    };

    render(
      <MockedProvider showWarnings={false} mocks={mocks2}>
        <Component {...variables2} />
      </MockedProvider>
    );

    waitFor(() => {
      expect(finished).toBe(true);
    }).then(resolve, reject);
  });

  itAsync('should not error if the variables match but have different order', (resolve, reject) => {
    let finished = false;
    function Component({ ...variables }: Variables) {
      const { loading, data } = useQuery<Data, Variables>(query, { variables });
      if (!loading) {
        expect(data).toMatchSnapshot();
        finished = true;
      }
      return null;
    }

    const mocks2 = [
      {
        request: {
          query,
          variables: {
            age: 13,
            username: 'some_user'
          }
        },
        result: { data: { user } }
      }
    ];

    const variables2 = {
      username: 'some_user',
      age: 13
    };

    render(
      <MockedProvider mocks={mocks2}>
        <Component {...variables2} />
      </MockedProvider>
    );

    waitFor(() => {
      expect(finished).toBe(true);
    }).then(resolve, reject);
  });

  itAsync('should support mocking a network error', (resolve, reject) => {
    let finished = false;
    function Component({ ...variables }: Variables) {
      const { loading, error } = useQuery<Data, Variables>(query, { variables });
      if (!loading) {
        expect(error).toEqual(new Error('something went wrong'));
        finished = true;
      }
      return null;
    }

    const mocksError = [
      {
        request: {
          query,
          variables
        },
        error: new Error('something went wrong')
      }
    ];

    render(
      <MockedProvider mocks={mocksError}>
        <Component {...variables} />
      </MockedProvider>
    );

    waitFor(() => {
      expect(finished).toBe(true);
    }).then(resolve, reject);
  });

  itAsync('should error if the query in the mock and component do not match', (resolve, reject) => {
    let finished = false;
    function Component({ ...variables }: Variables) {
      const { loading, error } = useQuery<Data, Variables>(query, { variables });
      if (!loading) {
        expect(error).toMatchSnapshot();
        finished = true;
      }
      return null;
    }

    const mocksDifferentQuery = [
      {
        request: {
          query: gql`
            query OtherQuery {
              otherQuery {
                id
              }
            }
          `,
          variables
        },
        result: { data: { user } }
      }
    ];

    render(
      <MockedProvider showWarnings={false} mocks={mocksDifferentQuery}>
        <Component {...variables} />
      </MockedProvider>
    );

    waitFor(() => {
      expect(finished).toBe(true);
    }).then(resolve, reject);
  });

  it('should pass down props prop in mock as props for the component', () => {
    function Component({ ...variables }) {
      expect(variables.foo).toBe('bar');
      expect(variables.baz).toBe('qux');
      return null;
    }

    render(
      <MockedProvider mocks={mocks} childProps={{ foo: 'bar', baz: 'qux' }}>
        <Component {...variables} />
      </MockedProvider>
    );
  });

  it('should not crash on unmount if there is no query manager', () => {
    function Component() {
      return null;
    }

    const { unmount } = render(
      <MockedProvider>
        <Component />
      </MockedProvider>
    );

    unmount();
  });

  itAsync('should support returning mocked results from a function', (resolve, reject) => {
    let finished = false;
    let resultReturned = false;

    const testUser = {
      __typename: 'User',
      id: 12345
    };

    function Component({ ...variables }: Variables) {
      const { loading, data } = useQuery<Data, Variables>(query, { variables });
      if (!loading) {
        expect(data!.user).toEqual(testUser);
        expect(resultReturned).toBe(true);
        finished = true;
      }
      return null;
    }

    const testQuery: DocumentNode = gql`
      query GetUser($username: String!) {
        user(username: $username) {
          id
        }
      }
    `;

    const testVariables = {
      username: 'jsmith'
    };
    const testMocks = [
      {
        request: {
          query: testQuery,
          variables: testVariables
        },
        result() {
          resultReturned = true;
          return {
            data: {
              user: {
                __typename: 'User',
                id: 12345
              }
            }
          };
        }
      }
    ];

    render(
      <MockedProvider mocks={testMocks}>
        <Component {...testVariables} />
      </MockedProvider>
    );

    waitFor(() => {
      expect(finished).toBe(true)
    }).then(resolve, reject);
  });

  it('should return "No more mocked responses" errors in response', async () => {
    let finished = false;
    function Component() {
      const { loading, error } = useQuery(query);
      if (!loading) {
        expect(error).toMatchSnapshot();
        finished = true;
      }
      return null;
    }

    const link = ApolloLink.from([
      errorLink,
      new MockLink([], true, { showWarnings: false })
    ]);

    render(
      <MockedProvider link={link}>
        <Component />
      </MockedProvider>
    );

    await waitFor(() => {
      expect(finished).toBe(true);
    });
    // The "No more mocked responses" error should not be thrown as an
    // uncaught exception.
    expect(errorThrown).toBeFalsy();
  });

  it('should return "Mocked response should contain" errors in response', async () => {
    let finished = false;
    function Component({ ...variables }: Variables) {
      const { loading, error } = useQuery<Data, Variables>(query, { variables });
      if (!loading) {
        expect(error).toMatchSnapshot();
        finished = true;
      }
      return null;
    }

    const link = ApolloLink.from([errorLink, new MockLink([
      {
        request: {
          query,
          variables
        }
      }
    ])]);

    render(
      <MockedProvider link={link}>
        <Component {...variables} />
      </MockedProvider>
    );

    await waitFor(() => {
      expect(finished).toBe(true);
    });
    // The "Mocked response should contain" error should not be thrown as an
    // uncaught exception.
    expect(errorThrown).toBeFalsy();
  });

  it('shows a warning in the console when there is no matched mock', async () => {
    const consoleSpy = jest.spyOn(console, 'warn').mockImplementation(() => {});
    let finished = false;
    function Component({ ...variables }: Variables) {
      const { loading } = useQuery<Data, Variables>(query, { variables });
      if (!loading) {
        finished = true;
      }
      return null;
    }

    const mocksDifferentQuery = [
      {
        request: {
          query: gql`
            query OtherQuery {
              otherQuery {
                id
              }
            }
          `,
          variables
        },
        result: { data: { user } }
      }
    ];

    render(
      <MockedProvider mocks={mocksDifferentQuery}>
        <Component {...variables} />
      </MockedProvider>
    );

    await waitFor(() => {
      expect(finished).toBe(true);
    });

    expect(console.warn).toHaveBeenCalledTimes(1);
    expect(console.warn).toHaveBeenCalledWith(
      expect.stringContaining('No more mocked responses for the query')
    );

    consoleSpy.mockRestore();
  });

  it('silences console warning for unmatched mocks when `showWarnings` is `false`', async () => {
    const consoleSpy = jest.spyOn(console, 'warn');
    let finished = false;
    function Component({ ...variables }: Variables) {
      const { loading } = useQuery<Data, Variables>(query, { variables });
      if (!loading) {
        finished = true;
      }
      return null;
    }

    const mocksDifferentQuery = [
      {
        request: {
          query: gql`
            query OtherQuery {
              otherQuery {
                id
              }
            }
          `,
          variables
        },
        result: { data: { user } }
      }
    ];

    render(
      <MockedProvider mocks={mocksDifferentQuery} showWarnings={false}>
        <Component {...variables} />
      </MockedProvider>
    );

    await waitFor(() => {
      expect(finished).toBe(true);
    });

    expect(console.warn).not.toHaveBeenCalled();

    consoleSpy.mockRestore();
  });

  it('silences console warning for unmatched mocks when passing `showWarnings` to `MockLink` directly', async () => {
    const consoleSpy = jest.spyOn(console, 'warn');
    let finished = false;
    function Component({ ...variables }: Variables) {
      const { loading } = useQuery<Data, Variables>(query, { variables });
      if (!loading) {
        finished = true;
      }
      return null;
    }

    const mocksDifferentQuery = [
      {
        request: {
          query: gql`
            query OtherQuery {
              otherQuery {
                id
              }
            }
          `,
         variables
        },
        result: { data: { user } }
      }
    ];

    const link = new MockLink(
      mocksDifferentQuery,
      false,
      { showWarnings: false }
    );

    render(
      <MockedProvider link={link}>
        <Component {...variables} />
      </MockedProvider>
    );

    await waitFor(() => {
      expect(finished).toBe(true);
    });

    expect(console.warn).not.toHaveBeenCalled();

    consoleSpy.mockRestore();
  });

  itAsync('should support custom error handling using setOnError', (resolve, reject) => {
    let finished = false;
    function Component({ ...variables }: Variables) {
      useQuery<Data, Variables>(query, { variables });
      return null;
    }

    const mockLink = new MockLink([], true, { showWarnings: false });
    mockLink.setOnError(error => {
      expect(error).toMatchSnapshot();
      finished = true;
    });
    const link = ApolloLink.from([errorLink, mockLink]);

    render(
      <MockedProvider link={link}>
        <Component {...variables} />
      </MockedProvider>
    );

    waitFor(() => {
      expect(finished).toBe(true);
    }).then(resolve, reject);
  });

  itAsync('should pipe exceptions thrown in custom onError functions through the link chain', (resolve, reject) => {
    let finished = false;
    function Component({ ...variables }: Variables) {
      const { loading, error } = useQuery<Data, Variables>(query, { variables });
      if (!loading) {
        expect(error).toMatchSnapshot();
        finished = true;
      }
      return null;
    }

    const mockLink = new MockLink([], true, { showWarnings: false });
    mockLink.setOnError(() => {
      throw new Error('oh no!');
    });
    const link = ApolloLink.from([errorLink, mockLink]);

    render(
      <MockedProvider link={link}>
        <Component {...variables} />
      </MockedProvider>
    );

    waitFor(() => {
      expect(finished).toBe(true);
    }).then(resolve, reject);
  });
});

describe('@client testing', () => {
  itAsync('should support @client fields with a custom cache', (resolve, reject) => {
    let finished = false;
    const cache = new InMemoryCache();

    cache.writeQuery({
      query: gql`{
        networkStatus {
          isOnline
        }
      }`,
      data: {
        networkStatus: {
          __typename: 'NetworkStatus',
          isOnline: true
        }
      },
    });

    function Component() {
      const { loading, data } = useQuery(gql`{
        networkStatus @client {
          isOnline
        }
      }`);
      if (!loading) {
        expect(data!.networkStatus.__typename).toEqual('NetworkStatus');
        expect(data!.networkStatus.isOnline).toEqual(true);
        finished = true;
      }
      return null;
    }

    render(
      <MockedProvider cache={cache}>
        <Component />
      </MockedProvider>
    );

    waitFor(() => {
      expect(finished).toBe(true);
    }).then(resolve, reject);
  });

  itAsync('should support @client fields with field policies', (resolve, reject) => {
    let finished = false;
    const cache = new InMemoryCache({
      typePolicies: {
        Query: {
          fields: {
            networkStatus() {
              return {
                __typename: 'NetworkStatus',
                isOnline: true
              };
            }
          }
        }
      }
    });

    function Component() {
      const { loading, data } = useQuery(gql`{
        networkStatus @client {
          isOnline
        }
      }`);
      if (!loading) {
        expect(data!.networkStatus.__typename).toEqual('NetworkStatus');
        expect(data!.networkStatus.isOnline).toEqual(true);
        finished = true;
      }
      return null;
    }

    render(
      <MockedProvider cache={cache}>
        <Component />
      </MockedProvider>
    );

    waitFor(() => {
      expect(finished).toBe(true);
    }).then(resolve, reject);
  });
});
