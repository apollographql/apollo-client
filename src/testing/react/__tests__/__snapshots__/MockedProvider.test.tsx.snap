// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`General use should allow querying with the typename 1`] = `
Object {
  "__typename": "User",
  "id": "user_id",
}
`;

exports[`General use should error if the query in the mock and component do not match 1`] = `
ApolloError {
  "cause": [Error: No more mocked responses for the query: query GetUser($username: String!) {
  user(username: $username) {
    id
    __typename
  }
}
Expected variables: {"username":"mock_username"}
],
  "clientErrors": Array [],
  "extraInfo": undefined,
  "graphQLErrors": Array [],
  "message": "No more mocked responses for the query: query GetUser($username: String!) {
  user(username: $username) {
    id
    __typename
  }
}
Expected variables: {\\"username\\":\\"mock_username\\"}
",
  "name": "ApolloError",
  "networkError": [Error: No more mocked responses for the query: query GetUser($username: String!) {
  user(username: $username) {
    id
    __typename
  }
}
Expected variables: {"username":"mock_username"}
],
  "protocolErrors": Array [],
}
`;

exports[`General use should error if the variableMatcher returns false 1`] = `
ApolloError {
  "cause": [Error: No more mocked responses for the query: query GetUser($username: String!) {
  user(username: $username) {
    id
    __typename
  }
}
Expected variables: {"username":"mock_username"}

Failed to match 1 mock for this query. The mocked response had the following variables:
  {}
],
  "clientErrors": Array [],
  "extraInfo": undefined,
  "graphQLErrors": Array [],
  "message": "No more mocked responses for the query: query GetUser($username: String!) {
  user(username: $username) {
    id
    __typename
  }
}
Expected variables: {\\"username\\":\\"mock_username\\"}

Failed to match 1 mock for this query. The mocked response had the following variables:
  {}
",
  "name": "ApolloError",
  "networkError": [Error: No more mocked responses for the query: query GetUser($username: String!) {
  user(username: $username) {
    id
    __typename
  }
}
Expected variables: {"username":"mock_username"}

Failed to match 1 mock for this query. The mocked response had the following variables:
  {}
],
  "protocolErrors": Array [],
}
`;

exports[`General use should error if the variables do not deep equal 1`] = `
ApolloError {
  "cause": [Error: No more mocked responses for the query: query GetUser($username: String!) {
  user(username: $username) {
    id
    __typename
  }
}
Expected variables: {"username":"some_user","age":42}

Failed to match 1 mock for this query. The mocked response had the following variables:
  {"age":13,"username":"some_user"}
],
  "clientErrors": Array [],
  "extraInfo": undefined,
  "graphQLErrors": Array [],
  "message": "No more mocked responses for the query: query GetUser($username: String!) {
  user(username: $username) {
    id
    __typename
  }
}
Expected variables: {\\"username\\":\\"some_user\\",\\"age\\":42}

Failed to match 1 mock for this query. The mocked response had the following variables:
  {\\"age\\":13,\\"username\\":\\"some_user\\"}
",
  "name": "ApolloError",
  "networkError": [Error: No more mocked responses for the query: query GetUser($username: String!) {
  user(username: $username) {
    id
    __typename
  }
}
Expected variables: {"username":"some_user","age":42}

Failed to match 1 mock for this query. The mocked response had the following variables:
  {"age":13,"username":"some_user"}
],
  "protocolErrors": Array [],
}
`;

exports[`General use should error if the variables in the mock and component do not match 1`] = `
ApolloError {
  "cause": [Error: No more mocked responses for the query: query GetUser($username: String!) {
  user(username: $username) {
    id
    __typename
  }
}
Expected variables: {"username":"other_user","age":<undefined>}

Failed to match 1 mock for this query. The mocked response had the following variables:
  {"username":"mock_username"}
],
  "clientErrors": Array [],
  "extraInfo": undefined,
  "graphQLErrors": Array [],
  "message": "No more mocked responses for the query: query GetUser($username: String!) {
  user(username: $username) {
    id
    __typename
  }
}
Expected variables: {\\"username\\":\\"other_user\\",\\"age\\":<undefined>}

Failed to match 1 mock for this query. The mocked response had the following variables:
  {\\"username\\":\\"mock_username\\"}
",
  "name": "ApolloError",
  "networkError": [Error: No more mocked responses for the query: query GetUser($username: String!) {
  user(username: $username) {
    id
    __typename
  }
}
Expected variables: {"username":"other_user","age":<undefined>}

Failed to match 1 mock for this query. The mocked response had the following variables:
  {"username":"mock_username"}
],
  "protocolErrors": Array [],
}
`;

exports[`General use should mock the data 1`] = `
Object {
  "__typename": "User",
  "id": "user_id",
}
`;

exports[`General use should not error if the variables match but have different order 1`] = `
Object {
  "user": Object {
    "__typename": "User",
    "id": "user_id",
  },
}
`;

exports[`General use should pipe exceptions thrown in custom onError functions through the link chain 1`] = `
ApolloError {
  "cause": [Error: oh no!],
  "clientErrors": Array [],
  "extraInfo": undefined,
  "graphQLErrors": Array [],
  "message": "oh no!",
  "name": "ApolloError",
  "networkError": [Error: oh no!],
  "protocolErrors": Array [],
}
`;

exports[`General use should return "Mocked response should contain" errors in response 1`] = `
ApolloError {
  "cause": [Error: Mocked response should contain either \`result\`, \`error\` or a \`delay\` of \`Infinity\`: {"query":"query GetUser($username: String!) {\\n  user(username: $username) {\\n    id\\n    __typename\\n  }\\n}"}],
  "clientErrors": Array [],
  "extraInfo": undefined,
  "graphQLErrors": Array [],
  "message": "Mocked response should contain either \`result\`, \`error\` or a \`delay\` of \`Infinity\`: {\\"query\\":\\"query GetUser($username: String!) {\\\\n  user(username: $username) {\\\\n    id\\\\n    __typename\\\\n  }\\\\n}\\"}",
  "name": "ApolloError",
  "networkError": [Error: Mocked response should contain either \`result\`, \`error\` or a \`delay\` of \`Infinity\`: {"query":"query GetUser($username: String!) {\\n  user(username: $username) {\\n    id\\n    __typename\\n  }\\n}"}],
  "protocolErrors": Array [],
}
`;

exports[`General use should return "No more mocked responses" errors in response 1`] = `
ApolloError {
  "cause": [Error: No more mocked responses for the query: query GetUser($username: String!) {
  user(username: $username) {
    id
    __typename
  }
}
Expected variables: {}
],
  "clientErrors": Array [],
  "extraInfo": undefined,
  "graphQLErrors": Array [],
  "message": "No more mocked responses for the query: query GetUser($username: String!) {
  user(username: $username) {
    id
    __typename
  }
}
Expected variables: {}
",
  "name": "ApolloError",
  "networkError": [Error: No more mocked responses for the query: query GetUser($username: String!) {
  user(username: $username) {
    id
    __typename
  }
}
Expected variables: {}
],
  "protocolErrors": Array [],
}
`;

exports[`General use should support custom error handling using setOnError 1`] = `
[Error: No more mocked responses for the query: query GetUser($username: String!) {
  user(username: $username) {
    id
    __typename
  }
}
Expected variables: {"username":"mock_username"}
]
`;

exports[`General use should use a mock if the variableMatcher returns true 1`] = `
Object {
  "__typename": "User",
  "id": "user_id",
}
`;
